pkg/deploy/lattice/target_group_manager.go

```diff
@@ -238,98 +237,98 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, modelTg *model.T
+	tgArnToTagsMap, err := s.cloud.Tagging().GetTagsFromArns(ctx, tgArns)
 
+	if err != nil {
+		return nil, err
+	}
 	for _, tg := range resp {
-		tgInput := vpclattice.GetTargetGroupInput{
-			TargetGroupIdentifier: tg.Id,
-		}
-
-		tgOutput, err := lattice.GetTargetGroupWithContext(ctx, &tgInput)
-		if err != nil {
-			continue
-		}
-
-		if tgOutput.Config != nil && aws.StringValue(tgOutput.Config.VpcIdentifier) == config.VpcID {
-			tagsInput := vpclattice.ListTagsForResourceInput{
-				ResourceArn: tg.Arn,
-			}
-
-			tagsOutput, err := lattice.ListTagsForResourceWithContext(ctx, &tagsInput)
-			if err != nil {
-				s.log.Infof("Failed ListTags %s: %s", aws.StringValue(tg.Arn), err)
-				// setting it to nil, so the caller knows this failed
-				tagsOutput = nil
-			}
-			tgOutput := tgListOutput{
-				getTargetGroupOutput: *tgOutput,
-				targetGroupTags:      tagsOutput,
-			}
-			tgList = append(tgList, tgOutput)
-		}
+		tgList = append(tgList, tgListOutput{
+			tgSummary: tg,
+			tags:      tgArnToTagsMap[*tg.Arn],
+		})
 	}
 	return tgList, err
 }
 
 func (s *defaultTargetGroupManager) findTargetGroup(
 	ctx context.Context,
 	modelTargetGroup *model.TargetGroup,
-) (*vpclattice.TargetGroupSummary, error) {
-	listInput := vpclattice.ListTargetGroupsInput{
-		VpcIdentifier: aws.String(modelTargetGroup.Spec.VpcId),
-	}
-	resp, err := s.cloud.Lattice().ListTargetGroupsAsList(ctx, &listInput)
+) (*vpclattice.GetTargetGroupOutput, error) {
+	arn, err := s.cloud.Tagging().FindResourceWithTags(ctx, services.ResourceTypeTargetGroup,
+		model.TagsFromTGTagFields(modelTargetGroup.Spec.TargetGroupTagFields))
 	if err != nil {
 		return nil, err
 	}
+	if arn == nil {
+		return nil, nil
+	}
 
-	for _, latticeTg := range resp {
-		// we ignore create failed status, so may as well check for it first
-		status := aws.StringValue(latticeTg.Status)
-		if status == vpclattice.TargetGroupStatusCreateFailed {
-			continue
+	latticeTg, err := s.cloud.Lattice().GetTargetGroupWithContext(ctx, &vpclattice.GetTargetGroupInput{
+		TargetGroupIdentifier: arn,
+	})
+	if err != nil {
+		if services.IsLatticeAPINotFoundErr(err) {
+			return nil, nil
 		}
+		return nil, err
+	}
 
-		isMatch, err := s.IsTargetGroupMatch(ctx, modelTargetGroup, latticeTg, nil)
-		if err != nil {
-			return nil, err
-		}
-		if isMatch {
-			s.log.Debugf("Target group %s already exists with arn %s", *latticeTg.Name, *latticeTg.Arn)
-			switch status {
-			case vpclattice.TargetGroupStatusCreateInProgress:
-				return nil, errors.New(LATTICE_RETRY)
-			case vpclattice.TargetGroupStatusActive:
-				return latticeTg, nil
-			case vpclattice.TargetGroupStatusDeleteFailed:
-				return latticeTg, nil
-			case vpclattice.TargetGroupStatusDeleteInProgress:
-				return nil, errors.New(LATTICE_RETRY)
-			}
+	// we ignore create failed status, so may as well check for it first
+	status := aws.StringValue(latticeTg.Status)
+	if status == vpclattice.TargetGroupStatusCreateFailed {
+		return nil, nil
+	}
```

Comment:
I dont follow why "findTargetGroup" make decisions about status and retry errors. It's a surprising behavior with "find" function. 

May be name is misleading? But it seems too much of responsibility for this function: searching, discarding "some" results, matching, retrying.

Reply:
Discarding some status should make total sense. It is always easier to think that `CREATE_FAILED` target group doesn't exist, and `DELETE_FAILED` target group just exists.

On the other hand, `LATTICE_RETRY` is currently like a magical fallback for us to "hold off" decision whenever we would like. It makes sense if you argue we are abusing it too much, but I find it aligning with other controller behaviors so didn't decide to touch yet.